File: examples/java/ExampleSaxpyBenchmark.java
import java.io.IOException;

import yacx.Executor;
import yacx.FloatArg;
import yacx.IntArg;
import yacx.KernelArg;
import yacx.Options;

public class ExampleSaxpyBenchmark {

	public static void main(String[] args) throws IOException {
		// Load library
		Executor.loadLibrary();

		// Benchmark saxpy-Kernel
		System.out.println(Executor.benchmark("saxpy", Options.createOptions(), 10, new Executor.KernelArgCreator<Integer>() {
			final float a = 5.1f;

			@Override
			public int getGrid0(Integer dataLength) {
				return dataLength;
			}

			@Override
			public int getBlock0(Integer dataLength) {
				return 1;
			}

			@Override
			public KernelArg[] createArgs(Integer dataLength) {
				float[] x = new float[dataLength];
				float[] y = new float[dataLength];

				for (int i = 0; i < dataLength; i++) {
					x[i] = 1;
					y[i] = i;
				}

				return new KernelArg[] { FloatArg.createValue(a), FloatArg.create(x), FloatArg.create(y),
						FloatArg.createOutput(dataLength), IntArg.createValue(dataLength) };
			}
		}, 1024, 4096, 8192, 1024 * 1024, 4096 * 1024, 16384 * 1024));
	}
}

File: examples/scala/ExampleSaxpyBenchmark.scala
import yacx.Executor;
import yacx.FloatArg;
import yacx.IntArg;
import yacx.KernelArg;
import yacx.Options;
import yacx.Utils;

object ExampleSaxpyBenchmark {
	private val KB = 1024
	
    def main(args: Array[String]) : Unit = {
    	//Load library
    	Executor.loadLibrary()
    	
      //Benchmark saxpy-Kernel
      println(Executor.benchmark("saxpy", Options.createOptions(), 10,
              new Executor.KernelArgCreator() {
                      val a = 5.1f
                			
          			      override def getDataLength(dataSizeBytes: Long) : Int = {
          						  return (dataSizeBytes/FloatArg.SIZE_BYTES).asInstanceOf[Int]
          					  }
          
          					  override def getGrid0(dataLength: Int) : Int = {
          						  return dataLength
          					  }
          
            					override def getBlock0(dataLength: Int) : Int = {
            						return 1
            					}
            
            					override def createArgs(dataLength: Int) : Array[KernelArg] = {
            						val x = new Array[Float](dataLength)
            						val y = new Array[Float](dataLength)
            
            						for (i <- 0 until dataLength) {
            							x(i) = 1
            							y(i) = i
            						}
            
            						return Array[KernelArg] (FloatArg.createValue(a), FloatArg.create(x: _*), FloatArg.create(y: _*),
            								FloatArg.createOutput(dataLength), IntArg.createValue(dataLength))
            					}
				}, 1*KB, 4*KB, 8*KB, 1024*KB, 4096*KB, 16384*KB))
    }
}